lib/nbd.c

```diff
@@ -0,0 +1,792 @@
+    return CURLE_AGAIN;
+
+  /* Otherwise try to receive. */
+  return nbd_recv(conn, msg_size, msg);
+}
+
+static CURLcode nbd_recv_magic(struct connectdata *conn)
+{
+  struct nbd_conn *nbdc = &conn->proto.nbdc;
+  struct handshake *handshake = NULL;
+  CURLcode result;
+
+  result = nbd_send_recv(conn, sizeof(struct handshake), (void **) &handshake);
+  if(result)
+    return result;
+  if(!handshake)
+    return CURLE_OK;
+
+  if(memcmp(handshake->nbdmagic, "NBDMAGIC", 8) != 0 ||
+     memcmp(handshake->version, "IHAVEOPT", 8) != 0) {
+    Curl_failf(conn->data,
+               "not an NBD server supporting the newstyle protocol");
+    return CURLE_WEIRD_SERVER_REPLY;
+  }
+
+  /* We only support "fixed newstyle".  All modern NBD servers should
+   * behave this way so if this fails it indicates some weird/old
+   * server.
+   */
+  nbdc->gflags = ntohs(handshake->gflags);
+  if((nbdc->gflags & NBD_FLAG_FIXED_NEWSTYLE) != NBD_FLAG_FIXED_NEWSTYLE) {
+    Curl_failf(conn->data,
+               "NBD server does not support the fixed newstyle protocol");
+    return CURLE_WEIRD_SERVER_REPLY;
+  }
+
+  /* Move to the next state. */
+  state(conn, NBD_SENDING_CLIENT_FLAGS);
+
+  return CURLE_OK;
+}
+
+static CURLcode nbd_send_client_flags(struct connectdata *conn)
+{
+  struct nbd_conn *nbdc = &conn->proto.nbdc;
+  unsigned int flags;
+  CURLcode result;
+
+  /* Send back only the global flags that we know about. */
+  flags = nbdc->gflags;
+  flags &= NBD_FLAG_FIXED_NEWSTYLE|NBD_FLAG_NO_ZEROES;
+  flags = htonl(flags);
+  result = nbd_send(conn, &flags, 4, 0);
+
+  /* Even though we haven't necessarily sent the data we can still
+   * move to the next state.
+   */
+  state(conn, NBD_SENDING_EXPORT_NAME);
+
+  return result;
+}
+
+static CURLcode nbd_send_export_name(struct connectdata *conn)
+{
+  struct nbd_conn *nbdc = &conn->proto.nbdc;
+  CURLcode result;
+  struct option *option;
+  size_t len = strlen(nbdc->exportname);
+  size_t msg_len = sizeof(struct option) + len - 1; /* because of data[1] */
+
+  option = malloc(msg_len);
+  if(option == NULL)
+    return CURLE_OUT_OF_MEMORY;
+  memcpy(option->version, "IHAVEOPT", 8);
+  option->option = htonl(NBD_OPT_EXPORT_NAME);
+  option->len = htonl((unsigned int)len);
+  memcpy(option->data, nbdc->exportname, len);
+  result = nbd_send(conn, option, msg_len, 0);
+  free(option);
+
+  /* Even though we haven't necessarily sent the data we can still
+   * move to the next state.
+   */
+  state(conn, NBD_EXPECTING_EXPORT_DATA);
+
+  return result;
+}
+
+static CURLcode nbd_recv_export_data(struct connectdata *conn)
+{
+  struct nbd_conn *nbdc = &conn->proto.nbdc;
+  struct export_data *data = NULL;
+  const bool uploading = conn->data->set.upload;
+  CURLcode result;
+
+  result = nbd_send_recv(conn, sizeof(struct export_data), (void **) &data);
+  if(result)
+    return result;
+  if(!data)
+    return CURLE_OK;
+
+  nbdc->size = be64toh(data->size);
```

Comment:
Is this (blindly) trusting that the remote server sends a valid size? What happens if it lies/is malicious?